README.md
# react-coordinate-input

> A masked input React component for entering latitude &amp; longitude coordinates as Degree Minute Second (DMS) values, with built-in conversion to Decimal Degrees.

[![NPM](https://flat.badgen.net/npm/v/react-coordinate-input)](https://www.npmjs.com/package/react-coordinate-input)
[![NPM](https://flat.badgen.net/npm/dt/react-coordinate-input)](https://www.npmjs.com/package/react-coordinate-input)
[![gzip size](https://flat.badgen.net/bundlephobia/minzip/react-coordinate-input)](https://bundlephobia.com/result?p=react-coordinate-input)
[![code coverage](https://flat.badgen.net/coveralls/c/github/nerdstep/react-coordinate-input)](https://coveralls.io/github/nerdstep/react-coordinate-input)
[![ci](https://github.com/nerdstep/react-coordinate-input/workflows/ci/badge.svg?branch=master)](https://github.com/nerdstep/react-coordinate-input/actions?query=workflow%3Aci)
[![license](https://flat.badgen.net/github/license/nerdstep/react-coordinate-input)](./LICENSE)

## Features

- Masked input with [imaskjs](https://github.com/uNmAnNeR/imaskjs)
- Coordinate validation
- Conversion to decimal degrees
- Accepts decimal degree passed in as `value` prop
- Second precision up to 6 decimal places
- Decimal degree precision up to 8 decimal places

## [Demo](https://nerdstep.github.io/react-coordinate-input/)

## Install

```bash
npm install --save react-coordinate-input
```

## Usage

```jsx
import React from 'react'
import CoordinateInput from 'react-coordinate-input'

const Example = () => {
  return (
    <CoordinateInput
      onChange={(value, { unmaskedValue, dd, dms }) => {
        console.log(value, unmaskedValue, dd, dms)
      }}
    />
  )
}
```

## Props

Component props and their default values.

```javascript
{
  // See below for more info on the onChange callback
  onChange: undefined,
  // Input placeholder; this will only be visible if `placeholderChar` is set to `null`
  placeholder: '04° 08′ 15″ N 162° 03′ 42″ E',
  // Number of decimal places to round decimal degrees to (0-8)
  ddPrecision: 6,
  // Number of decimal places for Seconds value on input (0-6)
  dmsPrecision: 0,
  // Callback function which receives the input ref
  inputRef: undefined,
  // DMS characters used in the input mask
  degreeChar: '°',
  minuteChar: '′',
  secondChar: '″',
  spacerChar: ' ',
  // The placeholder character represents the fillable spot in the input mask
  // Setting this to null will not display the mask guides
  placeholderChar: '_',
  // Input value -- see below for more info
  value: undefined,
}
```

### onChange

The `onChange` callback will only trigger once a valid coordinate has been entered or the input has been cleared.

The callback receives two arguments, the first being the masked input value, e.g. `90° 00′ 00″ N 180° 00′ 00″ W`, and the second is an object with three properties:

```javascript
{
  // The normalized DMS value
  unmaskedValue: '900000N1800000W',
  // The DMS value converted to Decimal Degrees
  dd: [90, -180],
  // DMS values split into an array of arrays,
  // e.g. [[D, M, S, 'N|S'], [D, M, S, 'E|W']]
  dms: [
    [90, 0, 0, 'N'], [180, 0, 0, 'W']
  ]
}
```

### Input value

When providing a value to the input the following formats are supported:

1. DMS value which conforms to the input mask, e.g. `04° 08′ 15″ N 162° 03′ 42″ E`

**Note:** The DMS symbols and spaces are not required; a minimal input would also work, e.g. `040815N1620342E`

The only requirement is that each value has the required number of characters, i.e. `04` not `4` for degrees, minutes, seconds, etc., except for longitude degrees which requires three characters.

2. Decimal degree string can also be used, e.g. `4.1375, 162.061667`

The component will detect the DD value and convert it to a DMS value automatically.

## Exports

In addition to the default `CoordinateInput` export, the following helper utilities are also exported for convenience:

**dmsToDecimal**

```js
/**
 * dmsToDecimal
 *
 * Converts Degrees Minutes Seconds to Decimal Degrees
 *
 * Formula:
 * DD = D + M / 60 + S / 3600
 *
 * @param {number} degrees
 * @param {number} minutes
 * @param {number} seconds
 * @param {string} direction Compass direction, e.g. N|S|E|W
 * @param {number} [precision]  Decimal places (default: 6)
 * @returns {number} Decimal degrees, e.g. 42.451
 */
```

**decimalToDMS**

```js
/**
 * decimalToDMS
 *
 * Converts Decimal Degress to Degrees Minutes Seconds
 *
 * @param {number} dd Decimal degree value
 * @param {boolean} isLon Is longitude?
 * @param {number} [precision] Decimal places for seconds (default: 0)
 * @returns {(string|number)[]} DMS values, e.g. [D, M, S, 'N|S|E|W']
 */
```

## License

MIT © [Justin M. Williams](https://github.com/nerdstep)


CHANGELOG.md
<a name="1.0.0"></a>

### 1.0.0 (2021-11-04)

#### Summary

- Moved `prop-types` from dependencies to `optionalDependencies`

<a name="1.0.0-rc.2"></a>

### 1.0.0-rc.2 (2020-11-04)

#### Summary

- This version refactors the internals to fix a bug where the input was not cleared after the component re-renders with an empty value.

#### Fixed

- The input is now cleared when an empty value is received.

<a name="1.0.0-rc.1"></a>

### 1.0.0-rc.1 (2020-10-29)

#### Summary

- This version fixes a little bug and refactors the component props a bit.

#### Changed

- Props such as `className`, `name`, etc. are no longer explicity defined as props on the component. Any props not specifically defined in the docs will be passed on to the input component.

#### Removed

- BREAKING CHANGE: Removed `inputProps` prop. Any props that you want to provide to the input can now simply be passed directly to the `CoordinateInput` component.

#### Fixed

- The `onChange` handler now correctly fires when the input has been cleared.

<a name="1.0.0-beta"></a>

### 1.0.0-beta (2020-05-15)

#### Summary

- This release is a pretty major refactor which replaces the unmaintained [Text Mask](https://github.com/text-mask/text-mask) library with [imaskjs](https://github.com/uNmAnNeR/imaskjs).
- Additionally, the component has been refactored to use React Hooks instead of Class Components.

#### Changed

- BREAKING CHANGE: The `onChange` callback signature has changed. Instead of receiving an event as the first argument it receives the masked input value. The second argument is still an object but the properties have changed. See the [docs](https://github.com/nerdstep/react-coordinate-input#onchange) for more info.

#### Removed

- BREAKING CHANGE: Removed `guide` prop
- BREAKING CHANGE: Removed `showMask` prop


example/src/App.js
import React, { useCallback, useEffect, useMemo, useRef, useState } from 'react'
import CoordinateInput from 'react-coordinate-input'
import Footer from './Footer'

function fill(arr, value, count) {
  for (let i = 0; i < count; i++) {
    arr.push(value)
  }
  return arr
}

const initialState = {
  value: '',
  unmaskedValue: '',
  dd: [],
  dms: [],
}

const selectOptions = [
  '90, -180',
  '-90, 180',
  '42.363, 27.891',
  '42 27 00 N 067 06 00 W',
  '040815N1620342E',
  '23° 16′ 07″ S 127° 00′ 00″ E',
]

const App = () => {
  const [ddPrecision, setDDPrecision] = useState(6)
  const [dmsPrecision, setDMSPrecision] = useState(0)
  const [selected, setSelected] = useState('')
  const [state, setState] = useState(initialState)
  const inputRef = useRef()

  const placeholder = useMemo(() => {
    const p = '04° 08′ 15.dec″ N 162° 03′ 42.dec″ E'
    let dec = ''
    if (dmsPrecision > 0) {
      dec = fill([], 0, dmsPrecision).join('')
      dec = `.${dec}`
    }
    return p.replace(/\.dec/g, dec)
  }, [dmsPrecision])

  const handleReset = useCallback(() => {
    setState(initialState)
    setSelected('')
  }, [])

  const handleChange = useCallback((value, props) => {
    console.log('handleChange', { value, ...props })
    setState({ value, ...props })
  }, [])

  const handleBlur = useCallback((e) => {
    //console.log('handleBlur', e)
  }, [])

  const handleChangeDDPrecision = useCallback(
    (e) => {
      handleReset()
      let value = parseInt(e.target.value, 10) || 0
      value = value < 0 ? 0 : value > 8 ? 8 : value
      setDDPrecision(value)
    },
    [handleReset]
  )

  const handleChangeDMSPrecision = useCallback(
    (e) => {
      handleReset()
      let value = parseInt(e.target.value, 10) || 0
      value = value < 0 ? 0 : value > 6 ? 6 : value
      setDMSPrecision(value)
    },
    [handleReset]
  )

  const handleSelectChange = useCallback(({ target: { value } }) => {
    setSelected(value)
    setState({ value })
  }, [])

  useEffect(() => {
    //console.log(inputRef.current.value)
  }, [])

  const output = {
    ...state,
    dms: JSON.stringify(state.dms),
    dd: JSON.stringify(state.dd),
  }

  return (
    <section className="hero is-fullheight has-background-light">
      <div className="hero-body">
        <div className="container">
          <div className="column is-8 is-offset-2">
            <h3 className="title has-text-grey has-text-centered">
              React Coordinate Input
            </h3>
            <div className="box">
              <div className="field">
                <label className="label">Input - Degrees Minutes Seconds</label>
                <div className="control">
                  <CoordinateInput
                    aria-label="coordinate-input"
                    className="input"
                    dir="auto"
                    ddPrecision={ddPrecision}
                    dmsPrecision={dmsPrecision}
                    inputRef={(c) => (inputRef.current = c)}
                    name="coordinate-input"
                    onBlur={handleBlur}
                    onChange={handleChange}
                    placeholder={placeholder}
                    placeholderChar="_"
                    value={state.value}
                  />
                </div>
                <p className="help">Example: {placeholder}</p>
              </div>
              <div className="field is-grouped">
                <div className="control">
                  <div className="select">
                    <select
                      disabled={dmsPrecision > 0}
                      onChange={handleSelectChange}
                      value={selected}
                    >
                      <option value="">Set value...</option>
                      {selectOptions.map((el, i) => (
                        <option key={i}>{el}</option>
                      ))}
                    </select>
                  </div>
                </div>
                <div className="control">
                  <button className="button" onClick={handleReset}>
                    Reset
                  </button>
                </div>
              </div>
              <h5>Options</h5>
              <div className="field">
                <label className="label is-small has-text-grey">
                  Decimal Places for Seconds (0-6)
                </label>
                <div className="control">
                  <input
                    className="input"
                    onChange={handleChangeDMSPrecision}
                    type="number"
                    value={dmsPrecision}
                  />
                </div>
              </div>
              <div className="field">
                <label className="label is-small has-text-grey">
                  Decimal Places for Decimal Degrees (0-8)
                </label>
                <div className="control">
                  <input
                    className="input"
                    onChange={handleChangeDDPrecision}
                    type="number"
                    value={ddPrecision}
                  />
                </div>
              </div>
              <div className="field">
                <label className="label">State</label>
                <div className="control">
                  <pre>{JSON.stringify(output, null, 2)}</pre>
                </div>
              </div>
            </div>
            <Footer />
          </div>
        </div>
      </div>
    </section>
  )
}

export default App
